File: backend/app/schemas/log.py
from datetime import datetime
from typing import Annotated, Literal, Optional

from pydantic import BaseModel, Field

SeverityType = Literal["INFO", "WARNING", "ERROR", "CRITICAL"]


class LogCreate(BaseModel):
    timestamp: Optional[datetime] = None
    severity: SeverityType
    source: Annotated[str, Field(min_length=1, max_length=255)]
    message: Annotated[str, Field(min_length=1)]


class LogUpdate(BaseModel):
    timestamp: Optional[datetime] = None
    severity: Optional[SeverityType] = None
    source: Optional[Annotated[str, Field(min_length=1, max_length=255)]] = None
    message: Optional[Annotated[str, Field(min_length=1)]] = None


class LogResponse(BaseModel):
    id: int
    timestamp: datetime
    severity: str
    source: str
    message: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class LogListResponse(BaseModel):
    data: list[LogResponse]
    total: int
    page: int
    limit: int
    pages: int


class SeveritySummary(BaseModel):
    INFO: int = 0
    WARNING: int = 0
    ERROR: int = 0
    CRITICAL: int = 0


class HistogramEntry(BaseModel):
    source: str
    INFO: int = 0
    WARNING: int = 0
    ERROR: int = 0
    CRITICAL: int = 0


class TimeseriesEntry(BaseModel):
    timestamp: datetime
    INFO: int = 0
    WARNING: int = 0
    ERROR: int = 0
    CRITICAL: int = 0


class AnalyticsSummaryResponse(BaseModel):
    summary: SeveritySummary
    histogram: list[HistogramEntry]


class TimeseriesResponse(BaseModel):
    interval: Literal["hour", "day", "week"]
    data: list[TimeseriesEntry]

File: frontend/src/types/log.ts
// Severity 型
export type Severity = 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL'

// Log エンティティ
export interface Log {
  id: number
  timestamp: string // ISO 8601
  severity: Severity
  source: string
  message: string
  created_at: string
  updated_at: string
}

// POST /logs リクエスト
export interface LogCreate {
  timestamp?: string
  severity: Severity
  source: string
  message: string
}

// PATCH /logs/{id} リクエスト（全フィールド省略可）
export interface LogUpdate {
  timestamp?: string
  severity?: Severity
  source?: string
  message?: string
}

// GET /logs レスポンス（ページネーション付き）
export interface LogListResponse {
  data: Log[]
  total: number
  page: number
  limit: number
  pages: number
}

// GET /logs クエリパラメータ
export interface LogListParams {
  start?: string
  end?: string
  severity?: Severity | Severity[]
  source?: string
  search?: string
  sort_by?: 'timestamp' | 'severity' | 'source'
  order?: 'asc' | 'desc'
  page?: number
  limit?: number
}

// GET /logs/analytics/summary レスポンス
export interface AnalyticsSummaryResponse {
  summary: {
    INFO: number
    WARNING: number
    ERROR: number
    CRITICAL: number
  }
  histogram: HistogramEntry[]
}

export interface HistogramEntry {
  source: string
  INFO: number
  WARNING: number
  ERROR: number
  CRITICAL: number
}

// GET /logs/analytics/timeseries レスポンス
export interface AnalyticsTimeseriesResponse {
  interval: 'hour' | 'day' | 'week'
  data: TimeseriesEntry[]
}

export interface TimeseriesEntry {
  timestamp: string
  INFO: number
  WARNING: number
  ERROR: number
  CRITICAL: number
}

// 分析系クエリパラメータ
export interface AnalyticsParams {
  start?: string
  end?: string
  severity?: Severity | Severity[]
  source?: string
  interval?: 'hour' | 'day' | 'week'
}

// CreateLogInput / UpdateLogInput (alias for clarity)
export type CreateLogInput = LogCreate
export type UpdateLogInput = LogUpdate
